airbnb/streamalert
# Create alert merger which dispatches alerts from Dynamo to alert processors

to: @ryandeivert 
cc: @airbnb/streamalert-maintainers 
size: large

## Background

This new Lambda function (the Alert Merger) will, as its name implies, eventually be responsible for alert merging. For now, it just dispatches alerts from the alerts Dynamo table and sends them to the alert processor.

The lifecycle of an alert is now as follows:

1. The rule processor generates alerts and saves them in batches to Dynamo
2. The alert merger runs every minute, finding all alerts in Dynamo which have not yet been sent or which need to be retried
3. The alert merger invokes the alert processor once for each alert, providing the Dynamo record as the payload
4. The alert processor deletes the Dynamo record once it is sent successfully, or else updates the list of outputs which need to be retried

Note that this PR does *not* implement alert merging, it just sets up the last piece of infra required to make it happen

## Changes

* Rule processor no longer invokes alert processor
* Add a new Alert Merger Lambda function, which runs every minute
* Update the alert processor with the new invocation format
* Support metric filters in the `tf_lambda` module
* Add a new `AlertAttempts` custom metric emitted by the Alert MErger

## Testing

* 100% coverage for the alert merger main
* Deploy to a test account and verify alerts end-to-end
* Deploy just the alert merger: `./manage.py lambda deploy -p alert_merger`
* Rollback just the alert merger: `./manage.py lambda rollback -p alert_merger`
* Try an alert which can't be sent and verify it is retried
* Enable and disable custom metrics
